addOns/replacer/src/test/java/org/zaproxy/zap/extension/replacer/ReplacerParamUnitTest.java

```diff
@@ -0,0 +1,191 @@
+
+    @Test
+    void shouldHaveConfigVersionKey() {
+        // Given / When
+        param.load(configuration);
+        // Then
+        assertThat(param.getConfigVersionKey(), is(equalTo("replacer[@version]")));
+    }
+
+    @ParameterizedTest
+    @ValueSource(strings = {"true", "false"})
+    void shouldLoadConfirmRemoveFromConfig(boolean state) {
+        // Given
+        configuration.setProperty("replacer.confirmRemoveToken", state);
+        // When
+        param.load(configuration);
+        // Then
+        assertThat(param.isConfirmRemoveToken(), is(equalTo(state)));
+    }
+
+    @Test
+    void shouldDefaultConfirmRemoveTrue() {
+        // Given / When
+        param.load(configuration);
+        // Then
+        assertThat(param.isConfirmRemoveToken(), is(equalTo(true)));
+    }
+
+    @Test
+    void shouldHaveExpectedVersionInDefaultConfig() {
+        // Given / When
+        param.load(configuration);
+        // Then
+        assertThat(param.getCurrentVersion(), is(equalTo(1)));
+    }
+
+    @Test
+    void shouldHaveExpectedNumberOfRulesOnUpdateFromUnversioned() {
+        // Given / When
+        param.load(configuration);
+        // Then
+        assertThat(param.getRules().size(), is(equalTo(EXPECTED_V1_RULE_COUNT)));
+    }
+
+    @Test
+    <T> void shouldNotOverwriteExistingRulesOnUpdate() {
+        // Given / When
+        param.load(configuration);
+        configuration = new ZapXmlConfiguration();
+        createConfigWithReportToRule(configuration);
+        param.load(configuration);
+        ReplacerParamRule enabledReportTo = getReportToRule(true);
+        param.addRule(enabledReportTo);
+        // Then
+        assertThat(param.getRules().size(), is(equalTo(2)));
+        // getRule returns first match (0th)
+        ReplacerParamRule originalRule = param.getRule(ReplacerParam.REPORT_TO_DESC);
+        assertNotNull(originalRule);
+        assertThat(originalRule.getDescription(), is(equalTo(ReplacerParam.REPORT_TO_DESC)));
+        assertThat(originalRule.isEnabled(), is(equalTo(false)));
+        // Added rule (1st) should be enabled
+        assertThat(param.getRules().get(1).isEnabled(), is(equalTo(true)));
+    }
+
+    @Test
+    void shouldNotOverwriteExistingRulesOnAdd() {
+        // Given / When
+        param.load(configuration);
+        ReplacerParamRule enabledReportTo = getReportToRule(true);
+        param.addRule(enabledReportTo);
+        // Then
+        assertThat(param.getRules().size(), is(equalTo(EXPECTED_V1_RULE_COUNT + 1)));
+        // getRule returns first match
+        ReplacerParamRule originalRule = param.getRule(ReplacerParam.REPORT_TO_DESC);
+        assertNotNull(originalRule);
+        assertThat(originalRule.getDescription(), is(equalTo(ReplacerParam.REPORT_TO_DESC)));
+        assertThat(originalRule.isEnabled(), is(equalTo(false)));
+        // Added rule (4th) should be enabled
+        assertThat(param.getRules().get(4).isEnabled(), is(equalTo(true)));
+    }
```

Comment:
This is failing for some reason. It comes up with 13 rules instead of 5. In isolation the test is fine, or if the call to                 saveRules();` is left in the `updateConfigImpl` before the fallthrough.

Reply:
@thc202 it seems that configuration is accumulating content despite being a supposedly new object each time. The only time it's in the expected state is when it's actually purposefully written to.

I added ` System.out.println(testinfo.getDisplayName() + " : " + param.getRules().size()); ` to all the tests which produces (I'm injecting `TestInfo testinfo` into each test as well):
```
[1] 2 : 2
[2] 999 : 2
shouldHaveExpectedVersionInDefaultConfig(TestInfo) : 12
[1] true : 16
[2] false : 20
shouldHaveExpectedRulesOnUpdateFromUnversioned(TestInfo) : 24
shouldDefaultConfirmRemoveTrue(TestInfo) : 28
shouldHaveConfigVersionKey(TestInfo) : 32
shouldNotOverwriteExistingRulesOnUpdate(TestInfo) : 2
```

The methods that output 2 have specifically done a cleartree on the rules and written to configuration.

Reply:
That's caused by the `setDefaultList()` call in the constructor, that really should be in a static block or better just inline the creation with `List.of(...)` when initialising the constant.

Reply:
Thank you!